pkg/reconciler/scheduling/namespace/namespace_reconcile_scheduling.go

```diff
@@ -0,0 +1,266 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package namespace
+
+import (
+	"context"
+	"encoding/json"
+	"math/rand"
+	"strings"
+	"time"
+
+	"github.com/kcp-dev/logicalcluster"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/klog/v2"
+	"k8s.io/kube-openapi/pkg/util/sets"
+
+	schedulingv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/scheduling/v1alpha1"
+	workloadv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/workload/v1alpha1"
+	locationreconciler "github.com/kcp-dev/kcp/pkg/reconciler/scheduling/location"
+)
+
+const removingGracePeriod = 10 * time.Second
+
+// placementSchedulingReconciler schedules a workload for this ns. It checks the current placement annotation on the ns,
+// and find all valid workloadclusters.
+type placementSchedulingReconciler struct {
+	listWorkloadCluster func(clusterName logicalcluster.Name) ([]*workloadv1alpha1.WorkloadCluster, error)
+
+	getPlacement func(clusterName logicalcluster.Name, name string) (*schedulingv1alpha1.Placement, error)
+
+	getLocation func(clusterName logicalcluster.Name, name string) (*schedulingv1alpha1.Location, error)
+
+	patchNamespace func(ctx context.Context, clusterName logicalcluster.Name, name string, pt types.PatchType, data []byte, opts metav1.PatchOptions, subresources ...string) (*corev1.Namespace, error)
+
+	enqueueAfter func(logicalcluster.Name, *corev1.Namespace, time.Duration)
+}
+
+func (r *placementSchedulingReconciler) reconcile(ctx context.Context, ns *corev1.Namespace) (reconcileStatus, *corev1.Namespace, error) {
+	clusterName := logicalcluster.From(ns)
+
+	var disableScheduling bool
+
+	if value, found := ns.Labels[workloadv1alpha1.SchedulingDisabledLabel]; found && value == "true" {
+		disableScheduling = true
+	}
+
+	var placementAnnotation schedulingv1alpha1.PlacementAnnotation
+	placementsValue, foundPlacement := ns.Annotations[schedulingv1alpha1.PlacementAnnotationKey]
+
+	// if scheduling is disabled, do not read placementAnnotation, so that all workloadclusters scheduled fot this ns
+	// will be marked as removing.
+	if !disableScheduling && foundPlacement {
+		if err := json.Unmarshal([]byte(placementsValue), &placementAnnotation); err != nil {
+			klog.Errorf("failed to unmarshal placement annotation %q of namespace %s|%s: %v", placementsValue, clusterName, ns.Name, err)
+		}
+	}
+
+	//1. pick all workload cluster in all bound placements
+	// it should be only one placement initially. But it could be multiple in the future for
+	// location federation.
```

Comment:
why not do it in the generic way directly?

Reply:
right, we do not need to treat specially